pymongo/change_stream.py

```diff
@@ -16,48 +16,125 @@
+    """The change stream cursor abstract base class.
 
-    .. versionadded: 3.6
-    .. mongodoc:: changeStreams
+    Defines the interface for change streams. Should be subclassed to
+    implement the `ChangeStream._create_cursor` abstract method and
+    the `ChangeStream._database` abstract property.
     """
-    def __init__(self, collection, pipeline, full_document,
-                 resume_after=None, max_await_time_ms=None, batch_size=None,
-                 collation=None, session=None):
-        self._collection = collection
+    def __init__(self, target, pipeline, full_document, resume_after,
+                 max_await_time_ms, batch_size, collation,
+                 start_at_operation_time, session):
+        if pipeline is None:
+            pipeline = []
+        elif not isinstance(pipeline, list):
+            raise TypeError("pipeline must be a list")
+
+        common.validate_string_or_none('full_document', full_document)
+        validate_collation_or_none(collation)
+        common.validate_non_negative_integer_or_none("batchSize", batch_size)
+
+        self._target = target
         self._pipeline = copy.deepcopy(pipeline)
         self._full_document = full_document
         self._resume_token = copy.deepcopy(resume_after)
         self._max_await_time_ms = max_await_time_ms
         self._batch_size = batch_size
         self._collation = collation
+        self._start_at_operation_time = start_at_operation_time
         self._session = session
         self._cursor = self._create_cursor()
 
-    def _full_pipeline(self):
+    @property
+    def _aggregation_target(self):
+        """The argument to pass to the aggregate command."""
+        raise NotImplementedError
+
+    @property
+    def _database(self):
+        """The database against which the aggregation commands for
+        this ChangeStream will be run. """
+        raise NotImplementedError
+
+    def _full_pipeline(self, inject_options=None):
         """Return the full aggregation pipeline for this ChangeStream."""
         options = {}
+
         if self._full_document is not None:
             options['fullDocument'] = self._full_document
         if self._resume_token is not None:
             options['resumeAfter'] = self._resume_token
+        if self._start_at_operation_time is not None:
+            options['startAtOperationTime'] = self._start_at_operation_time
+
+        if inject_options is not None:
+            options.update(inject_options)
         full_pipeline = [{'$changeStream': options}]
         full_pipeline.extend(self._pipeline)
         return full_pipeline
 
+    def _run_aggregation_cmd(self, session, explicit_session):
+        """Run the full aggregation pipeline for this ChangeStream and return
+        the corresponding CommandCursor.
+        """
+        with self._database.client._socket_for_reads(self._target._read_preference_for(session)) as (sock_info, slave_ok):
```

Comment:
Let's break this up to keep the lines under 80 chars.

read_preference = self._target._read_preference_for(session)
client = self._database.client
with client._socket_for_read(read_preference) as (sock_info, slave_ok):